hazelcast/src/hazelcast/client/compact.cpp

```diff
@@ -0,0 +1,719 @@
+                                      const std::string& field_name) const
+{
+    return { "compact_reader",
+             (boost::format("Unexpected fieldKind %1% for %2% on %3%") %
+              field_kind % field_name % schema)
+               .str() };
+}
+
+exception::hazelcast_serialization
+compact_reader::unexpected_null_value(const std::string& field_name,
+                                      const std::string& method_suffix)
+{
+    return { "compact_reader",
+             (boost::format(
+                "Error while reading %1%. null value can not be read via "
+                "get_%2% methods. Use get_nullable_%2%  instead.") %
+              field_name % method_suffix)
+               .str() };
+}
+
+size_t
+compact_reader::read_fixed_size_position(
+  const pimpl::field_descriptor& field_descriptor) const
+{
+    int primitive_offset = field_descriptor.offset();
+    return primitive_offset + data_start_position;
+}
+
+size_t
+compact_reader::read_var_size_position(
+  const pimpl::field_descriptor& field_descriptor) const
+{
+    int index = field_descriptor.index();
+    int offset =
+      get_offset(object_data_input, variable_offsets_position, index);
+    return offset == pimpl::offset_reader::NULL_OFFSET
+             ? pimpl::offset_reader::NULL_OFFSET
+             : offset + data_start_position;
+}
+
+int32_t
+compact_reader::read_int32(const std::string& field_name)
+{
+    const auto& fd = get_field_descriptor(field_name);
+    const auto& fieldKind = fd.field_kind();
+    switch (fieldKind) {
+        case pimpl::field_kind::INT32:
+            return object_data_input.read<int32_t>(
+              read_fixed_size_position(fd));
+        case pimpl::field_kind::NULLABLE_INT32:
+            return get_variable_as_non_null<int32_t>(fd, "int32");
+        default:
+            BOOST_THROW_EXCEPTION(unexpected_field_kind(fieldKind, field_name));
+            return -1;
+    }
+}
+
+int32_t
+compact_reader::read_int32(const std::string& field_name, int32_t default_value)
+{
+    return is_field_exists(field_name, pimpl::field_kind::INT32)
+             ? read_int32(field_name)
+             : default_value;
+}
+
+boost::optional<std::string>
+compact_reader::read_string(const std::string& field_name)
+{
+    return get_variable_size<std::string>(field_name,
+                                          pimpl::field_kind::STRING);
+}
+boost::optional<std::string>
+compact_reader::read_string(const std::string& field_name,
+                            const boost::optional<std::string>& default_value)
+{
+    return is_field_exists(field_name, pimpl::field_kind::STRING)
+             ? read_string(field_name)
+             : default_value;
+}
+
+namespace pimpl {
+
+compact_writer
+create_compact_writer(pimpl::default_compact_writer* default_compact_writer)
+{
+    return compact_writer{ default_compact_writer };
+}
+compact_writer
+create_compact_writer(pimpl::schema_writer* schema_writer)
```

Comment:
compact_writer constructor is private because it is a public API. But we need to be able to construct it internally in `compact_stream_serializer`. As a solution:
`create_compact_writer` is in pimpl and also friend method of `compact_writer`. So that we have a method to create compact-writer in `compact_stream_serializer`, and private to user(since it is in pimpl)